Initial WG LTS Reformation

jeremyrickard opened the pull request:
Initial PR to (re)create wg-lts. 

#7259 was opened as a tracker for this and has a lot of +1s. I also sent an email to the [dev mailing list](https://groups.google.com/a/kubernetes.io/g/dev/c/qWsMPMDqCFQ).  

This includes a draft charter that needs some eyes and reviews. 

Some things to determine:

- [ ] Identify a steering liaison
- [ ] Determine meeting times (follow up after chartering?)
- [ ] Agree on charter (part of this PR)


/committee steering

Adding stakeholder SIGs 

/sig architecture
/sig cluster-lifecycle
/sig k8s-infra
/sig release
/sig testing

SIG Approval

- [x] Architecture
- [x] Cluster Lifecycle
- [x] K8s Infra
- [x] Release
- [x] Testing
- [x] Security Response (committee)

/hold for consensus

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #7259

liggitt commented on sigs.yaml:
```diff
@@ -3245,6 +3245,44 @@ workinggroups:
+  - K8s Infra
+  - Release
+  - Testing
+  label: long-term-support
```
nit, `lts` everywhere is easier to remember (and matches all the things we used before, xref 0e9d714c7657733cebe94bb9e975cb7c121349c4)

xmudrii replied:
I actually wanted to leave the same comment but removed it before submitting the review. 😅 I'm not sure if `lts` is too short, but I general `lts` might be better.

jeremyrickard replied:
will do!

liggitt commented on wg-long-term-support/charter.md:
```diff
@@ -0,0 +1,81 @@
+
+## Stakeholders
+
+In the first phase of the working group, we will collect information related to the needs and wants regarding support periods from end-users and people supporting Kubernetes. This will be done by reaching out to users and cluster operator (e.g. via surveys), to better understand:
+
+* in-use versions (in the form of a user/vendor survey) and reasons for remaining on those versions
+* constraints on deployment and upgrade patterns / timelines (e.g. edge deployments, regulated industries, retail, etc)
+* expected/required support periods from users/vendors
+* support periods of core Kubernetes dependencies
+* support periods of other components required to run Kubernetes clusters (OS, network, storage, etc)
+* what users/vendors are currently doing to support releases past OSS End Of Life (EOL)
+* what do users/vendors do with EOL clusters today? what would they do if EOL was extended by N months/years?
+
+With this information, we will investigate changes the Kubernetes project could make to address these needs. A non-exhaustive set of changes to investigate include:
+* lengthening support period for all minor versions
+* lengthening support period for specific minor versions
+* applying security fixes to release branches past current EOL without cutting additional patch releases
+* applying security and "critical" fixes to release branches past current EOL without cutting additional patch releases
+* expanding supported skew
+* improving supported upgrade patterns for clusters at EOL
+
```
nit: might read easier if these are moved either to the scope or the deliverables section

liggitt reviewed (commented):
a couple naming/formatting suggestions, but 👍 on the overall shape

jeremyrickard commented on wg-long-term-support/charter.md:
```diff
@@ -0,0 +1,81 @@
+
+The artifacts the group is supposed to deliver include:
+- Survey results better describing Kubernetes version use patterns, deployment constraints, and upgrade patterns.
+- Recommend changes Kubernetes sigs can make that will provide broad benefits in a sustainable/affordable way. This would likely take the form of one or more KEPs.
```
thanks for catching that!

jeremyrickard commented on wg-long-term-support/charter.md:
```diff
@@ -0,0 +1,81 @@
+
+## Stakeholders
+
+In the first phase of the working group, we will collect information related to the needs and wants regarding support periods from end-users and people supporting Kubernetes. This will be done by reaching out to users and cluster operator (e.g. via surveys), to better understand:
```
thanks for catching that!

jeremyrickard commented:
Per Chairs/TL meeting: we should probably add all SIGs as stakeholders. @BenTheElder suggested a list of "owning" SIGS and "involved" SIGS and separate things out that way.  @mrbobbytables  mentioned we should almost certainly call out node and api-machinery as stakeholders.

jdumars commented:
Very glad to see this taking shape! Happy to approve when any objections/discussions are cleared.

ameukam commented:
+1 with my k8s-infra SIG chair hat.

ritazh commented on sigs.yaml:
```diff
@@ -3245,6 +3245,44 @@ workinggroups:
+  mission_statement: >
+    Define LTS
+
+  stakeholder_sigs:
```
IMO SRC should be a stakeholder here but we can deal with that later as part of the WG.

jeremyrickard replied:
I don't believe the generator/validation allows for anything other than valid SIGs here. I agree that committees could be stakeholders, but I don't think that has been a thin in the past? 

@kubernetes/steering-committee  what do you think should be done here?

jeremyrickard replied:
@ritazh I added the security response committee here as a label to capture that, for now.

jeremyrickard replied:
@palnabarun possibly, I was holding off adding any additional SIGs given the other discussions that have occurred but I will add this one because I think it's likely to be involved regardless.

ritazh commented on wg-lts/charter.md:
```diff
@@ -0,0 +1,81 @@
+
+## Roles and Organization Management
+
+This sig follows adheres to the Roles and Organization Management outlined in
```
```suggestion
This working group adheres to the Roles and Organization Management outlined in
```

youngnick commented on wg-lts/charter.md:
```diff
@@ -0,0 +1,81 @@
+
+## Timelines and Disbanding
+
+The exact timeline for this working group is hard to define at this time. If we are unable to define a common long term support definition for Kubernetes, we will disband the working group. If we are unable to define improvements related to an agreed upon definition, we will also disband the working group.
```
Should we write periodic updates and reviews to sig-arch/steering/someone else in here? Perhaps a 6-monthly update, and a review after 12 months?

Feels like having some deadlines to produce something might help keep the discussion focussed.

jeremyrickard replied:
@youngnick  that's a good suggestion, I will incorporate your feedback.

aojea commented:
+1 as Sig-testing TL

ritazh commented:
+1 on behalf of SRC

youngnick commented:
+1 on behalf of some guy who was involved last time.

dims commented on wg-lts/charter.md:
```diff
@@ -0,0 +1,81 @@
+* expanding supported skew
+* improving supported upgrade patterns for clusters at EOL
+
+### In Scope
```
Can you please consider some more of the items from the earlier WG charter? https://github.com/kubernetes/community/blob/master/archive/wg-lts/charter.md#in-scope

tpepper replied:
Are there any specific ones still missing at this point you'd like included?

dims commented on wg-lts/charter.md:
```diff
@@ -0,0 +1,81 @@
+- Creating and prioritizing a list of areas that require investments to improve long term supportability
+- Initiate and drive cross-SIG changes related to long term support
+
+### Out of scope
```
please consider adding at least the first bullet of the previous charter's out-of-scope section. https://github.com/kubernetes/community/blob/master/archive/wg-lts/charter.md#out-of-scope

palnabarun replied:
Resolving this since it's added.

dims commented:
One very critical piece this time around is there are more people in the community (especially vendors) who have public or private efforts (and people!) who help their own customers beyond community's efforts. How do we bring them to the table? (to do actual work!) Could we think of some things to get them engaged in the WG discussions from the get go and put those in the charter?

jeremyrickard commented:
Thanks for the feedback @dims. I added a new commit that brings in items from the previous working group charter and added two bullet points to the in scope section to address your last comment.

jeremyrickard commented:
Open question: should this PR leave the archive wg-lts for historical purposes or move the archive wg-lts out and apply the changes on top of that?

dims commented:
thanks @jeremyrickard 

> Open question: should this PR leave the archive wg-lts for historical purposes or move the archive wg-lts out and apply the changes on top of that?

+1 to leave it as-is please

pacoxu commented:
IMO, kubelet skew is one of the biggest problems for LTS. Have we asked SIG-Node to be one of the stakeholder SIGs?

pacoxu commented on wg-lts/charter.md:
```diff
@@ -0,0 +1,95 @@
+
+In the first phase of the working group, we will collect information related to the needs and wants regarding support periods from end-users and people supporting Kubernetes. This will be done by reaching out to users and cluster operators (e.g. via surveys), to better understand:
+
+* in-use versions (in the form of a user/vendor survey) and reasons for remaining on those versions
```
For the survey, we have done a survey of our customers that is similar.
- The survey should cover that users may have multi clusters in multi versions. The cluster size and number of clusters also matter for upgradation. For instance, migration is another approach for multi-clusters. Some big clusters may be hard to upgrade due to complex configurations and some special dependencies(storage/networking) of pods. Some customers chose to drain old-version clusters like draining a node. `Drain old version cluster` means to migrate apps to new clusters when apps need an update.
-  How often does your cluster upgrade minor versions? How often does your cluster upgrade patch versions?

ChaoyiHuang replied:
It's a little early to describe all survey questions in this initial reformation PR. Some examples are good enough.

Admit that survey questions will impact the WG-LTS conclusion greatly, need to be designed and reviewed carefully.

for exmple, if the question is expected support period for each minor version, may receive answer "the longer the better". But if the question is can you accept only one longer supported minor version every year, the answer maybe "yes".

Existing facts will be easier to answer and expressed exactly. Expectation is not equal to real need, sometimes will be overstated.

mrbobbytables replied:
Agreed - probably don't need to expand too much on survey questions here. It's fine to list a few as examples, but there will likely be a significant refinement of them. We do have some data already from the regular Production Readiness Surveys.

pacoxu commented on wg-lts/charter.md:
```diff
@@ -0,0 +1,95 @@
+* support periods of core Kubernetes dependencies
+* support periods of other components required to run Kubernetes clusters (OS, network, storage, etc)
+* what users/vendors are currently doing to support releases past community support End Of Life (EOL)
+* what do users/vendors do with EOL clusters today? what would they do if EOL was extended by N months/years?
```
> EOL was extended by N months/years.

Extending the current long-term support of all versions?

As I know, some other software will mark some versions as long support and others as not.
- for instance, we may mark one release of a year as long support for some extended time only. This may save some effort of the community.

liggitt replied:
That isn't known yet, that's one of the questions to investigate/answer

SergeyKanzhelev commented on sigs.yaml:
```diff
@@ -3248,6 +3248,44 @@ workinggroups:
+  mission_statement: >
+    Define LTS
+
+  stakeholder_sigs:
```
As @pacoxu already mentioned, let's put sig node here as well for the kubelet skew

liggitt replied:
Whether or not kubelet skew is relevant depends on the recommended changes, which are not yet known. I think ~every sig could be included as a stakeholder here, but I think it makes more sense to scope the sponsoring sigs to the ones impacted regardless of changes recommended

SergeyKanzhelev replied:
/lgtm

liggitt commented:
> kubelet skew is one of the biggest problems for LTS. Have we asked SIG-Node to be one of the stakeholder SIGs?

That depends on what is meant by LTS and what the proposed upgrade paths would be. I would probably wait to pull in sig node until there are specific proposals that would intersect with kubelet skew

cpanato commented:
/lgtm

jeremyrickard commented:
Thanks everyone for the reviews, comments, and discussion! 

@pacoxu raises some good points about the likely involvement of SIG Node, and really *every* SIG. I think that @liggitt is right though, that a lot will depend on what recommendations the WG actually makes and if/when we make recommendations, all appropriate SIGs will be stakeholders for those recommendations. 

I think we've had explicit approvals from the SIGs we did list as initial stakeholders, so I'm going to ping @kubernetes/steering-committee  for a review and comments now!

bharitwal commented on wg-lts/charter.md:
```diff
@@ -0,0 +1,95 @@
+
+## Timelines and Disbanding
+
+The exact timeline for this working group is hard to define at this time. If we are unable to define a common long term support definition for Kubernetes, we will disband the working group. If we are unable to define improvements related to an agreed upon definition, we will also disband the working group. In order to evaluate our progress toward the working group goals, we will provide periodic  ...
```
Would the lack of people and higher costs for providing long term support be a reason for disbanding the working group?

mrbobbytables replied:
I think that is more a potential outcome of the WG. If the WG determines we currently do not have the resources, nor commitments for future resources to potentially support an LTS then theres no reason to continue.

jeremyrickard replied:
I added this as reason for disbanding for clarity.

bharitwal reviewed (commented):
LGTM. Had a few questions.

mrbobbytables commented on wg-lts/charter.md:
```diff
@@ -0,0 +1,95 @@
+
+The artifacts the group is supposed to deliver include:
+- Survey results better describing Kubernetes version use patterns, deployment constraints, and upgrade patterns.
+- Recommend changes Kubernetes SIGS can make that will provide broad benefits in a sustainable/affordable way. This would likely take the form of one or more KEPs.
+- Recommend ways users/vendors who want to maximize Kubernetes support can consume Kubernetes
```
Generally if the target audience is end users, docs are pushed to the primary website + blog article highlighting them.

For disruptive changes we do engage other communities more directly - e.g. pushing to reddit, CNCF end user group etc.

mrbobbytables reviewed (commented):
No major blockers, but a few small comments.

tpepper commented on sigs.yaml:
```diff
@@ -3248,6 +3248,44 @@ workinggroups:
       name: Christoph Blecker
+- dir: wg-lts
+  name: LTS
+  mission_statement: >
+    Define LTS
```
I like the idea of moving the start of the scope section here.  It reads too ambiguous as it stands.  Sure the goal is (updated?) definition for long term support in the Kubernetes project.  But for first pass reforming this needs spelled out more for the reader.

Then in the coming months if the WG bootstraps and refines the mission based on initial discussions and data gathering, that can follow in subsequent updates if not already covered by a generic/sufficient starting statement.

jeremyrickard replied:
Thanks for the suggestion, I will add this.

palnabarun commented on wg-lts/charter.md:
```diff
@@ -0,0 +1,95 @@
+
+The artifacts the group is supposed to deliver include:
+- Survey results better describing Kubernetes version use patterns, deployment constraints, and upgrade patterns.
+- Recommend changes Kubernetes SIGS can make that will provide broad benefits in a sustainable/affordable way. This would likely take the form of one or more KEPs.
```
nit:
```suggestion
- Recommend changes Kubernetes SIGs can make that will provide broad benefits in a sustainable/affordable way. This would likely take the form of one or more KEPs.
```

palnabarun commented:
@jeremyrickard -- can you PTAL at the comments and also rebase the branch?

jeremyrickard commented on sigs.yaml:
```diff
@@ -3248,6 +3248,44 @@ workinggroups:
+    recordings_url: TBD
+  contact:
+    slack: wg-lts
+    mailing_list: https://groups.google.com/forum/#!forum/kubernetes-wg-lts
```
Good point.

jeremyrickard replied:
I'll mark it as TBD for now and update later if this is approved and we have created the new group.

jeremyrickard commented on wg-lts/charter.md:
```diff
@@ -0,0 +1,95 @@
+  * Costs of Kubernetes releases in terms of:
+    * Infrastructure
+    * People
+- Initiate and drive cross-SIG changes related to long term support.
+- Identify ways to that vendors and other community members can better support Kubernetes releases.
```
Thanks!

jeremyrickard commented:
@palnabarun  I don't know if we have a core set of dependencies documented. If we don't, as part of the WG we can document those and capture their support periods in order to better understand this space.  I reworded the sentence a little to indicate that.

jeremyrickard commented:
Thanks to @mrbobbytables @tpepper @palnabarun @bharitwal for the most recent round of comments! 

I've rebased this PR and tried to incorporate your feedback! Please take a look again :)

Notable changes:

* Added SIG Security per @palnabarun 
* Updated mission statement in the sigs.yaml file per everyone. 
* Minor updates for other suggestions.

bharitwal commented on wg-lts/README.md:
```diff
@@ -0,0 +1,40 @@
+--->
+# LTS Working Group
+
+The working group is organized with the goal of developing a better understanding of what "Long Term Support" might mean for Kubernetes, those who support  Kubernetes, and end users. The working group will investigate changes the Kubernetes project could make related this better understanding of what long term support might mean. The working group will also determine the feasibility, benefits, ** ...
```
```suggestion
The working group is organized with the goal of developing a better understanding of what "Long Term Support" might mean for Kubernetes, those who support Kubernetes, and end users. The working group will investigate changes the Kubernetes project could make related to the better understanding of what long term support might mean. The working group will also determine the feasibility, benefits, **cost**, and prerequisites  of any such changes.
```

bharitwal commented:
> @jeremyrickard: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-community-verify	[48f671a](https://github.com/kubernetes/community/commit/48f671af3c0dbd4864d8186477015fdb3637b429)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/community/7287/pull-community-verify/1673431166352363520)	true	`/test pull-community-verify`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=community&pr=7287). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Ajeremyrickard). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/community/issues?q=is:issue+is:open) when you hit one in your PR.

@jeremyrickard I see the tests are failing as `Readme.md` is an auto-generated file and sigs.yaml should have been updated.

jeremyrickard commented:
> > @jeremyrickard: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> > Test name	Commit	Details	Required	Rerun command
> > pull-community-verify	[48f671a](https://github.com/kubernetes/community/commit/48f671af3c0dbd4864d8186477015fdb3637b429)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/community/7287/pull-community-verify/1673431166352363520)	true	`/test pull-community-verify`
> > [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=community&pr=7287). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Ajeremyrickard). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/community/issues?q=is:issue+is:open) when you hit one in your PR.
> 
> @jeremyrickard Apologies, I see tests are failing as `Readme.md` is an auto-generated file and sigs.yaml should have been updated.

Yeah, sorry I didn't catch that either :) I'll rebase and include your suggestion in the sigs.yaml file

jeremyrickard commented:
@kubernetes/steering-committee  ping for another review, thanks!

cpanato commented:
/lgtm

tpepper reviewed (commented):
lgtm (Steering)

palnabarun commented on sigs.yaml:
```diff
@@ -3212,6 +3212,50 @@ workinggroups:
+  contact:
+    slack: wg-lts
+    mailing_list: TBD
+    liaison:
+      github: TBD
+      name: TBD
```
I volunteer to be the liaison. (unless someone else wants to be the liaison)

cc: @kubernetes/steering-committee

palnabarun replied:
SGTM either way.

palnabarun reviewed (approved):
/lgtm
(with the steering hat on)

palnabarun commented:
Steering Approvals:
- [ ] @BenTheElder 
- [ ] @cblecker  
- [x] @cpanato 
- [ ] @justaugustus 
- [x] @mrbobbytables
- [x] @palnabarun 
- [x] Tim Pepper

MadhavJivrajani commented on wg-lts/charter.md:
```diff
@@ -0,0 +1,97 @@
+
+Any changes identified by this working group will require involvement and investment from the following SIGS:
+
+- SIG Architecture
+  High-level input on requirements.
+- SIG Cluster Lifecycle
+  Input on cluster upgrade mechanics.
+- SIG K8s Infra
+  Input on infrastructure costs related to increased support period(s).
+- SIG Release
+  Input on maintaining older branches and additional releases.
+- SIG Testing
+  Input on testing impacts related to upgrades and longer test periods. 
```
I see that SIG Security was added as a stakeholder, maybe an update here slipped through the cracks?

jeremyrickard replied:
Thanks @MadhavJivrajani, let me fix that up.

cblecker commented on sigs.yaml:
```diff
@@ -3212,6 +3212,50 @@ workinggroups:
+- dir: wg-lts
+  name: LTS
+  mission_statement: >
+    The working group is organized with the goal of developing a better understanding
+    of what "Long Term Support" might mean for Kubernetes, those who support  Kubernetes,
+    and end users. The working group will investigate changes the Kubernetes project
+    could make related to the better understanding of what long term support might
+    mean. The working group will also determine the feasibility, benefits, **cost**,
+    and prerequisites  of any such changes.
```
```suggestion
    The working group is organized with the goal of developing a better understanding
    of what "Long Term Support" might mean for Kubernetes, those who support Kubernetes,
    and end users. The working group will investigate changes the Kubernetes project
    could make related to the better understanding of what long term support might
    mean. The working group will also determine the feasibility, benefits, **cost**,
    and prerequisites of any such changes.
```

cblecker reviewed (commented):
non-blocking formatting nit, otherwise lgtm.

wzshiming commented on wg-lts/README.md:
```diff
@@ -0,0 +1,40 @@
+--->
+# LTS Working Group
+
+The working group is organized with the goal of developing a better understanding of what "Long Term Support" might mean for Kubernetes, those who support  Kubernetes, and end users. The working group will investigate changes the Kubernetes project could make related to the better understanding of what long term support might mean. The working group will also determine the feasibility, benefits,  ...
```
```suggestion
The working group is organized with the goal of developing a better understanding of what "Long Term Support" might mean for Kubernetes, those who support Kubernetes, and end users. The working group will investigate changes the Kubernetes project could make related to the better understanding of what long term support might mean. The working group will also determine the feasibility, benefits, **cost**, and prerequisites of any such changes.
```

wzshiming commented on wg-lts/charter.md:
```diff
@@ -0,0 +1,99 @@
+- The lifecycle of projects outside of the Kubernetes org.
+- Designing and executing on changes clearly falling into individual SIG
+  responsibilities. This is a working group, no code implementation is the responsibility of this Working Group.
+- Technical and end-user support:  The WG may make recommendations
```
```suggestion
- Technical and end-user support: The WG may make recommendations
```

jeremyrickard commented:
Looks like we now have approval from the following steering committee members:

- [ ] @BenTheElder
- [x] @cblecker
- [x] @cpanato
- [ ] @justaugustus
- [x] @mrbobbytables
- [x] @palnabarun
- [x] @tpepper

I believe this constitutes a simple majority? Are we ok to drop the hold on this and merge it? It will need another LGTM/Approve since I rolled in the suggestion from @cblecker. If possible, I would like to merge this and begin the work of spawning a mailing list and slack channels, and follow up to this with additional data as we create it.

cblecker commented:
@jeremyrickard Let me check once more with Stephen and Ben. Yes, we have a majority but if they plan on reviewing we prefer to have full consensus for something as important as a new community group.

jeremyrickard commented:
> @jeremyrickard Let me check once more with Stephen and Ben. Yes, we have a majority but if they plan on reviewing we prefer to have full consensus for something as important as a new community group.

Thanks @cblecker, that makes sense!

BenTheElder reviewed (commented):
From a Steering Committee Perspective: LGTM. 

Sorry for the delay, this has been high on my list between vacations.

With my SIG Testing and Kubernetes maintainer hats on: I personally have reservations about the sustainability etc, but the workgroup will determine possible directions and then we can evaluate feasibility etc as a community :-)

jdumars commented:
Hi Community - based on the feedback, it appears this is ready to LGTM and remove the hold, which I will do by EOD barring any objections. I echo Ben's concerns and hope that we can mobilize around this effort and make it sustainable. Thanks everyone for your support on this! 

/lgtm
